Code/GraphMol/RascalMCES/RascalMCES.cpp

```diff
@@ -0,0 +1,1088 @@
+
+// Make the set of pairs of vertices, where they're a pair if the labels match.
+void buildPairs(const ROMol &mol1, const std::vector<unsigned int> &vtxLabels1,
+                const ROMol &mol2, const std::vector<unsigned int> &vtxLabels2,
+                const RascalOptions &opts,
+                std::vector<std::pair<int, int>> &vtxPairs) {
+  std::vector<std::string> mol1RingSmiles, mol2RingSmiles;
+  std::vector<std::unique_ptr<ROMol>> mol1Rings, mol2Rings;
+  // For these purposes, it is correct that n1cccc1 and [nH]1cccc1 match - the
+  // former would be from an N-substituted pyrrole, the latter from a plain one.
+  static const std::regex reg(R"(\[([np])H\])");
+  if (opts.completeAromaticRings) {
+    extractRings(mol1, mol1Rings, mol1RingSmiles);
+    for (auto &mrs : mol1RingSmiles) {
+      mrs = std::regex_replace(mrs, reg, "$1");
+    }
+    extractRings(mol2, mol2Rings, mol2RingSmiles);
+    for (auto &mrs : mol2RingSmiles) {
+      mrs = std::regex_replace(mrs, reg, "$1");
+    }
+  }
+
+  for (auto i = 0u; i < vtxLabels1.size(); ++i) {
+    for (auto j = 0u; j < vtxLabels2.size(); ++j) {
+      if (vtxLabels1[i] == vtxLabels2[j]) {
+        if (opts.completeAromaticRings &&
+            !checkAromaticRings(mol1, mol1RingSmiles, i, mol2, mol2RingSmiles,
+                                j)) {
+          continue;
+        }
+        if (opts.ringMatchesRingOnly &&
+            !checkRingMatchesRing(mol1, i, mol2, j)) {
+          continue;
+        }
+        vtxPairs.push_back(std::make_pair(i, j));
+      }
+    }
+  }
+}
+
+// Make the modular product between the 2 graphs passed in.  Each node in the
+// graph is a pair of vertices, one from the first graph, the other from the
+// second, whose labels match.  Two vertices are connected in the modular
+// product if either the 2 matching vertices in the 2 input vertices are
+// connected by edges with the same label, or neither is connected.
+void makeModularProduct(const ROMol &mol1,
+                        const std::vector<std::vector<int>> &adjMatrix1,
+                        const std::vector<unsigned int> &vtxLabels1,
+                        const std::vector<std::vector<int>> &distMatrix1,
+                        const ROMol &mol2,
+                        const std::vector<std::vector<int>> &adjMatrix2,
+                        const std::vector<unsigned int> &vtxLabels2,
+                        const std::vector<std::vector<int>> &distMatrix2,
+                        const RascalOptions &opts,
+                        std::vector<std::pair<int, int>> &vtxPairs,
+                        std::vector<std::vector<char>> &modProd) {
+  buildPairs(mol1, vtxLabels1, mol2, vtxLabels2, opts, vtxPairs);
+  if (vtxPairs.empty()) {
+    // There was nothing in common at all.  But, what was the screening doing?
+    modProd.clear();
+    return;
+  }
+  if (vtxPairs.size() > opts.maxBondMatchPairs) {
+    std::cerr << "Too many matching bond pairs (" << vtxPairs.size()
```

Comment:
this should use the logging mechanism and go to either the warning log or the info log.